docs/contrib/keycloak.md

```diff
@@ -0,0 +1,148 @@
+# Keycloak Identity Provider for REMS
+
+Using docker, docker-compose, and Keycloak, you can roll your own local IdP for use with REMS.
+
+## Quick-ish Start
+
+1. **Setting up the Docker files**
+    1. Create a `Dockerfile` for your keycloak. To start, it can be as simple as the following:
+    ```
+    FROM quay.io/keycloak/keycloak:11.0.2
+    ```
+    2. Create a `docker-compose.yaml` file for REMS and your Keycloak IdP. Mount your custom REMS config to `/rems/config/config.edn`. Example:
+    ```
+    version: "3.7"
+    services:
+    rems:
+        mem_reservation: 200m
+        mem_limit: 500m
+        ports:
+        - "127.0.0.1:3001:3000"
+        volumes:
+        - ./services/rems/simple-config.edn:/rems/config/config.edn
+        depends_on:
+        - db
+        - keycloak
+    db:
+        image: postgres:9.6
+        environment:
+        POSTGRES_USER: rems
+        POSTGRES_PASSWORD: remspassword
+        mem_reservation: 30m
+        mem_limit: 150m
+        ports:
+        - "127.0.0.1:5432:5432"
+    keycloak:
+        environment:
+        KEYCLOAK_USER: admin
+        KEYCLOAK_PASSWORD: admin
+        ports:
+        - "3002:8080"
+    ```
+
+2. **Preparing a Keycloak Realm**
+    1. Build and run keycloak with `docker-compose up keycloak`.
```

Comment:
I had some difficulty with these separate ups but I think there is a bit of a chicken & egg going on with Keycloak needing to be configured before REMS container can start because it goes looking for the auth configs.

Reply:
Maybe Keycloak can be configured completely with some initial files? Like having that export?

Reply:
Keycloak definitely needs to be up before REMS, because REMS tries to connect with it on start-up. There is not *too* much configuration that needs to be done with Keycloak once you do all of these steps once, and then export the realm with its clients. Subsequent configuration just looks like:
1. Spin up keycloak with `docker-compose up keycloak`
2. Log into the admin console
3. Import realm
4. (optional) Add a user.

You can even import/export users somehow, and probably make initialization even faster, but I haven't succeeded in doing this yet myself. I will definitely come back and add docs on how to do that once I figure it out.

From [these keycloak standalone docs](https://github.com/keycloak/keycloak-documentation/blob/master/server_admin/topics/export-import.adoc) it looks like it may be possible to do the realm import from a CLI. But since this constitutes a migration I don't think it would be run as a stage in `docker run`/`docker-compose up`.

Reply:
If you would like a sample `realm-export.json` file for serving to REMS, I would be happy to provide one.

Reply:
I managed with your instructions but a sample realm with the same settings as an option sounds nice. We don't need to make this perfect now, there are probably more pressing issues to work on. If I was using a lot of Keycloak myself, I would probably try to build an image that contains all the configuration too.